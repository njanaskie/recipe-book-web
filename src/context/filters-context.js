import React, { useContext, useReducer } from 'react';
import filtersReducer, { filtersReducerDefaultState } from '../reducers/filters'

const FiltersContext = React.createContext()

export const useFiltersContext = () => useContext(FiltersContext)

const FiltersProvider = ({ children }) => {
    const [filters, filtersDispatch] = useReducer(filtersReducer, filtersReducerDefaultState)

    const filterIngredients = (ingredients) => {
        filtersDispatch({ type: 'SET_INGREDIENT_FILTER', ingredients })
    }

    const filterCuisine = (cuisine) => {
        filtersDispatch({ type: 'SET_CUISINE_FILTER', cuisine })
    }

    const filterType = (recipeType) => {
        filtersDispatch({ type: 'SET_RECIPE_TYPE_FILTER', recipeType })
    }

    const filterCustomTags = (customTags) => {
        filtersDispatch({ type: 'SET_CUSTOM_TAG_FILTER', customTags })
    }

    return (
        <FiltersContext.Provider value={{ filters, filtersDispatch, filterCuisine, filterIngredients, filterType, filterCustomTags }}>
            {children}
        </FiltersContext.Provider>
    )
}

export { FiltersProvider, FiltersContext as default }