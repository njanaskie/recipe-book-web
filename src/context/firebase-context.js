import React, { useEffect, useState, useContext, useReducer } from 'react';
import database, { firebase, googleAuthProvider } from '../firebase/firebase';
import { history } from '../routers/AppRouter';
import authReducer from '../reducers/auth'
import LoadingPage from '../components/LoadingPage'

const FirebaseContext = React.createContext()

export const useFirebaseContext = () => useContext(FirebaseContext)

const FirebaseProvider = ({ children }) => {
    const [auth, authDispatch] = useReducer(authReducer, [])
    const [user, setUser] = useState('')
    const [loading, setLoading] = useState(true)
    const [isAdmin, setIsAdmin] = useState()
    const [isGuest, setIsGuest] = useState()
    const email = process.env.GUEST_EMAIL
    const password = process.env.GUEST_PASSWORD

    const login = () => {
        firebase.auth().signInWithPopup(googleAuthProvider).then(() => {
            authDispatch({ type: 'LOGIN' })
        })
    }

    const loginAsGuest = () => {
        firebase.auth().signInWithEmailAndPassword(email, password).then(() => {
            authDispatch({ type: 'LOGIN_AS_GUEST' })
        })
    }

    const logout = () => {
        firebase.auth().signOut().then(() => {
            authDispatch({ type: 'LOGOUT' })
        });
    }

    useEffect(() => {
        const unsubscribe = firebase.auth().onAuthStateChanged((user) => {
            setUser(user)
            setLoading(false)
            setIsAdmin()
            setIsGuest()
        })

        return () => unsubscribe()
    }, [])

    useEffect(() => {
        if (user) {
            database.collection('users').doc(user.uid).get()
            .then((doc) => {
                const userData = doc.data()
                if (userData) {
                    if (userData.isAdmin === true) {
                        const isAdmin = userData.isAdmin
                        setIsAdmin(isAdmin)
                    } else if (userData.isGuest === true) {
                        const isGuest = userData.isGuest
                        setIsGuest(isGuest)
                    }
                }
            })
        }
    }, [user])

    if (loading) {
        return <LoadingPage />
    }
    // if (user) {
    //     if (history.location.pathname === '/') {
    //         history.push('/home')
    //     }
    // } else {
    //     history.push('/');
    // }

    return (
        <FirebaseContext.Provider value={{ user, loading, isAdmin, isGuest, login, loginAsGuest, logout }}>
            {children}
        </FirebaseContext.Provider>
    )
}

export { FirebaseProvider, FirebaseContext as default }