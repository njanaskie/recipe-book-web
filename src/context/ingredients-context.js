import React, { useContext, useReducer } from 'react';
import ingredientsReducer from '../reducers/ingredients'
import database from '../firebase/firebase'

const IngredientsContext = React.createContext()

export const useIngredientsContext = () => useContext(IngredientsContext)

const IngredientsProvider = ({ children }) => {
    const [ingredients, dispatch ] = useReducer(ingredientsReducer, [])

    const removeIngredient = ({ id } = {}) => {
        return database.collection('ingredients').doc(id).delete().then(() => {
            dispatch({ type: 'REMOVE_INGREDIENT', id })
        });
    };

    const addIngredient = (ingredient) => {
        database.collection('ingredients').add(ingredient).then((ref) => {
            dispatch(({ type: 'ADD_INGREDIENT', ingredient: {id: ref.key, ...ingredient} }))
        })
    }

    return (
        <IngredientsContext.Provider value={{ ingredients, dispatch, removeIngredient, addIngredient }}>
            {children}
        </IngredientsContext.Provider>
    )
}

export { IngredientsProvider, IngredientsContext as default }