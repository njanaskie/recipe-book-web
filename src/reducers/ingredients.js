
// Ingredients Reducer

const ingredientsReducerDefaultState = [];

const ingredientsReducer = (state = ingredientsReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_INGREDIENT':
            return [
                ...state,
                action.ingredient
            ];
        case 'REMOVE_INGREDIENT':
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_INGREDIENT':
            return state.map((ingredient) => {
                if (ingredient.id === action.id) {
                    return {
                        ...ingredient,
                        ...action.updates
                    };
                } else {
                    return ingredient;
                };
            });
        case 'SET_INGREDIENTS':
            return action.ingredients;
        default:
            return state;
    }
}

export default ingredientsReducer;