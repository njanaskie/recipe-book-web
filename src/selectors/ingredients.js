
export default (ingredients) => {
    return ingredients
    .sort((a, b) => {
        if (a.category === b.category) {
            return b.category - a.category
        }
         return a.category > b.category ? 1 : -1
    })
}