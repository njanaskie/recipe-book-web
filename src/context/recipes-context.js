import React, { useContext, useReducer } from 'react';
import recipesReducer from '../reducers/recipes'
import database from '../firebase/firebase'
import uuid from 'uuid'
import { useFirebaseContext } from './firebase-context'


const RecipesContext = React.createContext()

export const useRecipesContext = () => useContext(RecipesContext)

const RecipesProvider = ({ children }) => {
    const { user } = useFirebaseContext() 
    const [recipes, recipeDispatch] = useReducer(recipesReducer, [])
    const customId = uuid()

    const addRecipe = (recipe) => {
        database.collection('users').doc(user.uid).collection('recipes').doc(customId).set(recipe).then(() => {
            recipeDispatch({ type: 'ADD_RECIPE', recipe: {id: customId, ...recipe} })
        })
    }

    const editRecipe = (id, updates) => {
        database.collection('users').doc(user.uid).collection('recipes').doc(id).update(updates).then(() => {
            recipeDispatch({ type: 'EDIT_RECIPE', id, updates })
        })
    }

    const removeRecipe = (id) => {
        database.collection('users').doc(user.uid).collection('recipes').doc(id).delete().then(() => {
            recipeDispatch({ type: 'REMOVE_RECIPE', id })
        })
    }

    return (
        <RecipesContext.Provider value={{ recipes, recipeDispatch, addRecipe, editRecipe, removeRecipe }}>
            {children}
        </RecipesContext.Provider>
    )
}

export { RecipesProvider, RecipesContext as default }