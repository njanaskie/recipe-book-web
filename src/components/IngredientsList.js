import React  from 'react'
import { List } from 'semantic-ui-react'
import IngredientListItem from './IngredientListItem'
import { useIngredientsContext } from '../context/ingredients-context'
import selectIngredients from '../selectors/ingredients'

export const IngredientsList = () => {
    const { ingredients } = useIngredientsContext()
    const selectedIngredients = selectIngredients(ingredients)

    if (!selectedIngredients || !selectedIngredients.length) {
        return <div className='content-container'><span className="list-item--message">No ingredients</span></div>
    }

    const tableItems = selectedIngredients.map((ingredient, id) => {
        return (
            <IngredientListItem
                key={id}
                ingredient={ingredient}
            />
        )
    })

    return (
        <div className="content-container">
            <List divided verticalAlign='middle'>
                {tableItems}
            </List>
        </div>
    )
}

export { IngredientsList as default }