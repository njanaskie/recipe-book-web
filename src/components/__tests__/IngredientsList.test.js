import React from 'react'
import { shallow, mount } from 'enzyme'
import { IngredientsList } from '../IngredientsList'
import ingredients from '../../../tests/fixtures/ingredients'
import * as AllIngredientsContext from '../../../context/ingredients-context'
import IngredientsContext from '../../../context/ingredients-context'
import * as AllPantryContext from '../../../context/pantry-context'
import PantryContext from '../../../context/pantry-context'
import * as AllFirebaseContext from '../../../context/firebase-context'
import FirebaseContext from '../../../context/firebase-context'
import user from '../../../tests/fixtures/user'

let userSpy, pantryIngredientsSpy

beforeEach(() => {
    userSpy = jest.spyOn(AllFirebaseContext, 'useFirebaseContext').mockImplementation(() => user)
    pantryIngredientsSpy = jest.spyOn(AllPantryContext, 'usePantryContext').mockImplementation(() => [])
})

afterEach(() => {
    userSpy.mockRestore()
    pantryIngredientsSpy.mockRestore()
})

test('should render IngredientsList with ingredients', () => {
    const wrapper = shallow(
        <IngredientsContext.Provider value={ingredients}>
            <IngredientsList ingredients={ingredients}/>
        </IngredientsContext.Provider>
    )
    expect(wrapper).toMatchSnapshot()
})

test('should render IngredientsList with empty message',() => {
    jest.spyOn(AllIngredientsContext, 'useIngredientsContext').mockImplementation(() => [])
    const wrapper = shallow(
        <IngredientsContext.Provider value={{ ingredients }}>
            <PantryContext.Provider value={{ pantryIngredients: pantryIngredientsSpy }}>
                <IngredientsList ingredients={[]}/>
            </PantryContext.Provider>
        </IngredientsContext.Provider>
    ).dive().dive()
    expect(wrapper).toMatchSnapshot()
});

// test('should render IngredientsList with ingredients2',() => {
//     jest.spyOn(AllIngredientsContext, 'useIngredientsContext').mockImplementation(() => ingredients)
//     const wrapper = mount(
//         <IngredientsContext.Provider value={{ ingredients }}>
//             <PantryContext.Provider value={{ pantryIngredients: pantryIngredientsSpy }}>
//                 <FirebaseContext.Provider value={{ user: userSpy }}>
//                     <IngredientsList ingredients={ingredients}/>
//                 </FirebaseContext.Provider>
//             </PantryContext.Provider>
//         </IngredientsContext.Provider>, { context: ingredients }
//     ).dive().dive().dive()
//     expect(wrapper).toMatchSnapshot()
// });