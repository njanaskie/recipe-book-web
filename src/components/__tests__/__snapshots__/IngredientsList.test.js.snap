// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render IngredientsList with empty message 1`] = `
<div>
  <div>
    <span>
      No ingredients
    </span>
  </div>
</div>
`;

exports[`should render IngredientsList with ingredients 1`] = `
<IngredientsList
  ingredients={
    Array [
      Object {
        "category": "Spices",
        "id": "1",
        "name": "Salt",
        "price": 1,
      },
      Object {
        "category": "Vegetables",
        "id": "2",
        "name": "Lettuce",
        "price": 1,
      },
      Object {
        "category": "Dairy",
        "id": "3",
        "name": "Yoghurt",
        "price": 2,
      },
    ]
  }
/>
`;
